File: ios/TrainLibreLiveActivity/RecoveryWidget.swift
import SwiftUI
import WidgetKit

// MARK: - Timeline

@available(iOS 18.0, *)
struct RecoveryEntry: TimelineEntry {
  let date: Date
  let recovery: HomeWidgetRecovery?
}

@available(iOS 18.0, *)
struct RecoveryProvider: TimelineProvider {
  func placeholder(in context: Context) -> RecoveryEntry {
    RecoveryEntry(date: .now, recovery: .placeholder)
  }

  func getSnapshot(in context: Context, completion: @escaping (RecoveryEntry) -> Void) {
    let stored = HomeWidgetSnapshot.load()?.recovery
    completion(RecoveryEntry(date: .now, recovery: stored ?? (context.isPreview ? .placeholder : nil)))
  }

  /// One entry, refreshed daily.
  ///
  /// Recovery does change while the app is closed — muscles keep recovering on
  /// the clock — but the state boundaries are 36 to 72 hours out, so a widget
  /// that re-renders once a day cannot be more than a few hours behind, and the
  /// app reloads this timeline whenever a workout is finished. Polling for the
  /// exact hour a muscle crosses from "recovering" to "ready" would spend the
  /// whole daily reload budget on a distinction nobody trains by.
  func getTimeline(in context: Context, completion: @escaping (Timeline<RecoveryEntry>) -> Void) {
    let entry = RecoveryEntry(date: .now, recovery: HomeWidgetSnapshot.load()?.recovery)
    completion(Timeline(entries: [entry], policy: .after(.now.addingTimeInterval(6 * 60 * 60))))
  }
}

// MARK: - Widget

@available(iOS 18.0, *)
struct RecoveryWidget: Widget {
  var body: some WidgetConfiguration {
    StaticConfiguration(
      kind: TrainLibreHomeWidget.kindRecovery,
      provider: RecoveryProvider()
    ) { entry in
      RecoveryWidgetView(recovery: entry.recovery)
        .statsWidgetContainer()
        .widgetURL(URL(string: "trainlibre://analytics/recovery"))
    }
    .configurationDisplayName(Text("widget.recovery.name"))
    .description(Text("widget.recovery.description"))
    .supportedFamilies([.systemMedium])
    .contentMarginsDisabled()
  }
}

// MARK: - View

@available(iOS 18.0, *)
struct RecoveryWidgetView: View {
  let recovery: HomeWidgetRecovery?

  @Environment(\.colorScheme) private var colorScheme
  @Environment(\.widgetRenderingMode) private var renderingMode

  private var palette: StatsPalette {
    StatsPalette(colorScheme: colorScheme, renderingMode: renderingMode)
  }

  private var hasData: Bool {
    guard let recovery else { return false }
    return recovery.hasData && !recovery.states.isEmpty
  }

  var body: some View {
    VStack(alignment: .leading, spacing: 0) {
      StatsHeader(
        title: Text("widget.recovery.name"),
        chip: hasData ? Text("widget.recovery.chip") : nil,
        palette: palette
      )

      Text(headline)
        .font(StatsTheme.headlineFont)
        .foregroundStyle(headlineColor)
        .lineLimit(2)
        .minimumScaleFactor(0.85)
        .fixedSize(horizontal: false, vertical: true)
        .padding(.top, 4)

      Spacer(minLength: 8)

      if hasData, let states = recovery?.states {
        HStack(spacing: StatsTheme.rowSpacing) {
          ForEach(Array(states.enumerated()), id: \.offset) { index, state in
            RecoveryPill(
              state: state,
              palette: palette,
              // Recovering is the state that wants attention, fresh the one
              // that does not. In tinted mode this ordering is the only thing
              // left to tell them apart.
              emphasis: emphasis(forIndex: index, of: states.count)
            )
          }
        }
      } else {
        emptyBody
      }
    }
  }

  private var headline: String {
    guard let recovery, !recovery.headline.isEmpty else { return "" }
    return recovery.headline
  }

  private var headlineColor: Color {
    guard hasData else { return palette.secondaryText }
    return palette.stateColor(recovery?.headlineColorHex)
  }

  private func emphasis(forIndex index: Int, of count: Int) -> Double {
    guard count > 1 else { return 1 }
    return 1 - Double(index) / Double(count - 1)
  }

  /// No workout has been logged yet, so there is nothing to recover from.
  private var emptyBody: some View {
    VStack(alignment: .leading, spacing: 9) {
      Text("widget.recovery.empty.body")
        .font(.system(size: 12))
        .foregroundStyle(palette.secondaryText)
        .lineLimit(2)
        .fixedSize(horizontal: false, vertical: true)

      StatsActionPill(
        label: Text("widget.startWorkout"),
        destination: URL(string: "trainlibre://action/start_workout")!,
        palette: palette
      )
    }
    .frame(maxWidth: .infinity, alignment: .leading)
    .padding(12)
    .background(palette.secondarySurface, in: RoundedRectangle(cornerRadius: StatsTheme.pillRadius, style: .continuous))
    .overlay(
      RoundedRectangle(cornerRadius: StatsTheme.pillRadius, style: .continuous)
        .stroke(Color.primary.opacity(0.07), lineWidth: 1)
    )
  }
}

/// One of the three readiness counts — count, label, share.
@available(iOS 18.0, *)
struct RecoveryPill: View {
  let state: HomeWidgetRecoveryState
  let palette: StatsPalette
  let emphasis: Double

  var body: some View {
    VStack(alignment: .leading, spacing: 3) {
      Text("\(state.count)")
        .font(.system(size: 16, weight: .bold))
        .foregroundStyle(palette.stateColor(state.colorHex))
        .lineLimit(1)

      Text(state.label)
        .font(.system(size: 14, weight: .medium))
        .foregroundStyle(palette.onSurface)
        .lineLimit(1)
        .minimumScaleFactor(0.7)

      Text("\(state.percent)%")
        .font(.system(size: 12))
        .foregroundStyle(palette.secondaryText)
        .lineLimit(1)
    }
    .frame(maxWidth: .infinity, alignment: .leading)
    .padding(.horizontal, 12)
    .padding(.vertical, 10)
    .background(palette.secondarySurface, in: RoundedRectangle(cornerRadius: StatsTheme.pillRadius, style: .continuous))
    .overlay(
      RoundedRectangle(cornerRadius: StatsTheme.pillRadius, style: .continuous)
        .stroke(
          palette.stateBorder(state.colorHex, emphasis: emphasis),
          lineWidth: palette.stateBorderWidth(emphasis: emphasis)
        )
    )
  }
}

// MARK: - Gallery placeholder

@available(iOS 18.0, *)
extension HomeWidgetRecovery {
  /// What the widget gallery shows, and what a widget added before the app has
  /// ever written a snapshot falls back to. A plausible mid-week distribution
  /// rather than an error state.
  static var placeholder: HomeWidgetRecovery {
    HomeWidgetRecovery(
      hasData: true,
      headline: String(localized: "widget.recovery.placeholder.headline"),
      headlineColorHex: "#FF9800",
      states: [
        HomeWidgetRecoveryState(
          state: "recovering",
          label: String(localized: "widget.recovery.state.recovering"),
          count: 6, percent: 46, colorHex: "#FF9800"
        ),
        HomeWidgetRecoveryState(
          state: "ready",
          label: String(localized: "widget.recovery.state.ready"),
          count: 2, percent: 15, colorHex: "#2196F3"
        ),
        HomeWidgetRecoveryState(
          state: "fresh",
          label: String(localized: "widget.recovery.state.fresh"),
          count: 5, percent: 38, colorHex: "#4CAF50"
        ),
      ]
    )
  }
}

File: android/app/src/main/kotlin/com/rfivesix/trainlibre/widgets/RecoveryWidget.kt
package com.rfivesix.trainlibre.widgets

import androidx.compose.runtime.Composable
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.glance.GlanceModifier
import androidx.glance.LocalContext
import androidx.glance.appwidget.GlanceAppWidget
import androidx.glance.appwidget.GlanceAppWidgetReceiver
import androidx.glance.layout.Alignment
import androidx.glance.layout.Box
import androidx.glance.layout.Column
import androidx.glance.layout.Row
import androidx.glance.layout.Spacer
import androidx.glance.layout.fillMaxWidth
import androidx.glance.layout.height
import androidx.glance.layout.padding
import androidx.glance.layout.width
import androidx.glance.text.FontWeight
import androidx.glance.text.Text
import androidx.glance.text.TextStyle
import com.rfivesix.trainlibre.R
import com.rfivesix.trainlibre.widgets.glance.SnapshotWidget
import com.rfivesix.trainlibre.widgets.glance.openOnTap
import com.rfivesix.trainlibre.widgets.snapshot.HomeWidgetRecovery
import com.rfivesix.trainlibre.widgets.snapshot.HomeWidgetRecoveryState
import com.rfivesix.trainlibre.widgets.snapshot.HomeWidgetSnapshot
import com.rfivesix.trainlibre.widgets.theme.RoundedShape
import com.rfivesix.trainlibre.widgets.theme.StatsActionPill
import com.rfivesix.trainlibre.widgets.theme.StatsHeader
import com.rfivesix.trainlibre.widgets.theme.StatsPalette
import com.rfivesix.trainlibre.widgets.theme.StatsTheme
import com.rfivesix.trainlibre.widgets.theme.provider
import com.rfivesix.trainlibre.widgets.theme.roundedBackground
import com.rfivesix.trainlibre.widgets.theme.statsWidgetContainer

/**
 * The muscle readiness card.
 *
 * Port of `RecoveryWidget` in `ios/TrainLibreLiveActivity/RecoveryWidget.swift`.
 * Composed rather than drawn: the pills are text on a bordered surface, which
 * Glance can express, and keeping them as real text means they scale with the
 * user's font size and are readable by TalkBack.
 */
class RecoveryWidget : SnapshotWidget() {

    @Composable
    override fun Content(snapshot: HomeWidgetSnapshot?) {
        val context = LocalContext.current
        val palette = StatsPalette.of(context)
        val recovery = snapshot?.recovery
        val hasData = recovery != null && recovery.hasData && recovery.states.isNotEmpty()

        Column(
            modifier = GlanceModifier
                .statsWidgetContainer(palette)
                .openOnTap(WidgetDeepLinks.RECOVERY),
        ) {
            StatsHeader(
                title = context.getString(R.string.widget_recovery_name),
                palette = palette,
                chip = if (hasData) context.getString(R.string.widget_recovery_chip) else null,
            )
            Spacer(GlanceModifier.height(4.dp))

            Text(
                text = recovery?.headline.orEmpty(),
                maxLines = 2,
                style = StatsTheme.headlineStyle.copy(
                    color = if (hasData) {
                        palette.stateColor(recovery?.headlineColorHex).provider()
                    } else {
                        palette.secondaryText.provider()
                    },
                ),
            )

            Spacer(GlanceModifier.defaultWeight())

            if (hasData && recovery != null) {
                States(recovery, palette)
            } else {
                EmptyBody(palette)
            }
        }
    }

    @Composable
    private fun States(recovery: HomeWidgetRecovery, palette: StatsPalette) {
        Row(modifier = GlanceModifier.fillMaxWidth()) {
            recovery.states.forEachIndexed { index, state ->
                if (index > 0) Spacer(GlanceModifier.width(StatsTheme.rowSpacing))
                Box(modifier = GlanceModifier.defaultWeight()) {
                    StatePill(state, palette)
                }
            }
        }
    }

    /**
     * One of the three readiness counts — count, label, share.
     *
     * Two nested boxes rather than one bordered shape: a `ColorFilter` tints a
     * whole drawable, so a fill and a border that need different colours cannot
     * share one. The outer box is the ring, the inner one covers all but a
     * hairline of it.
     */
    @Composable
    private fun StatePill(state: HomeWidgetRecoveryState, palette: StatsPalette) {
        Box(
            modifier = GlanceModifier
                .fillMaxWidth()
                .roundedBackground(palette.stateBorder(state.colorHex), RoundedShape.PillStroke)
                .padding(1.dp),
        ) {
            Column(
                modifier = GlanceModifier
                    .fillMaxWidth()
                    .roundedBackground(palette.secondarySurface, RoundedShape.Pill)
                    .padding(horizontal = 12.dp, vertical = 10.dp),
            ) {
                Text(
                    text = state.count.toString(),
                    maxLines = 1,
                    style = TextStyle(
                        fontSize = 16.sp,
                        fontWeight = FontWeight.Bold,
                        color = palette.stateColor(state.colorHex).provider(),
                    ),
                )
                Spacer(GlanceModifier.height(3.dp))
                Text(
                    text = state.label,
                    maxLines = 1,
                    style = TextStyle(
                        fontSize = 14.sp,
                        fontWeight = FontWeight.Medium,
                        color = palette.onSurface.provider(),
                    ),
                )
                Spacer(GlanceModifier.height(3.dp))
                Text(
                    text = "${state.percent}%",
                    maxLines = 1,
                    style = TextStyle(fontSize = 12.sp, color = palette.secondaryText.provider()),
                )
            }
        }
    }

    /** No workout has been logged yet, so there is nothing to recover from. */
    @Composable
    private fun EmptyBody(palette: StatsPalette) {
        val context = LocalContext.current
        Column(
            modifier = GlanceModifier
                .fillMaxWidth()
                .roundedBackground(palette.secondarySurface, RoundedShape.Pill)
                .padding(12.dp),
            horizontalAlignment = Alignment.Start,
        ) {
            Text(
                text = context.getString(R.string.widget_recovery_empty_body),
                maxLines = 2,
                style = TextStyle(fontSize = 12.sp, color = palette.secondaryText.provider()),
            )
            Spacer(GlanceModifier.height(9.dp))
            StatsActionPill(
                label = context.getString(R.string.widget_start_workout),
                deepLink = WidgetDeepLinks.action(QuickAction.StartWorkout.key),
                palette = palette,
            )
        }
    }
}

class RecoveryWidgetReceiver : GlanceAppWidgetReceiver() {
    override val glanceAppWidget: GlanceAppWidget = RecoveryWidget()
}
